File: library/src/main/java/me/proxer/library/entity/notifications/Notification.java
package me.proxer.library.entity.notifications;

import com.squareup.moshi.Json;
import lombok.Getter;
import lombok.Value;
import me.proxer.library.entity.ProxerDateItem;
import me.proxer.library.entity.ProxerIdItem;
import me.proxer.library.enums.NotificationType;
import okhttp3.HttpUrl;

import javax.annotation.Nullable;
import java.util.Date;

/**
 * @author Ruben Gees
 */
@Value
public class Notification implements ProxerIdItem, ProxerDateItem {

    /**
     * Returns the id.
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "id")
    private String id;

    /**
     * Returns the type of the notification.
     */
    @Json(name = "type")
    private NotificationType type;

    /**
     * Returns the id of the content. This depends on the {@link #type}.
     */
    @Json(name = "tid")
    private String contentId;

    /**
     * Returns the link to the content.
     */
    @Json(name = "link")
    private HttpUrl contentLink;

    /**
     * Returns the description of the content. This is directly usable for presentation.
     */
    @Json(name = "linktext")
    private String text;

    /**
     * Returns the time.
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "time")
    private Date date;

    /**
     * Returns additional info, depending on the {@link #type}.
     */
    @Getter(onMethod = @__({@Nullable}))
    @Json(name = "description")
    private String additionalDescription;
}

File: library/src/main/kotlin/me/proxer/library/entity/notifications/Notification.kt
package me.proxer.library.entity.notifications

import com.squareup.moshi.Json
import me.proxer.library.entity.ProxerDateItem
import me.proxer.library.entity.ProxerIdItem
import me.proxer.library.enums.NotificationType
import okhttp3.HttpUrl
import java.util.Date

/**
 * Entity representing a single notification.
 * This can be for example a new anime episode or a new forum topic.
 *
 * @property type The type of the notification.
 * @property contentId The id of the content. This depends on the [type].
 * @property contentLink The link to the content.
 * @property text The description of the content. This is directly usable for presentation.
 * @property additionalDescription Additional info, depending on the [type].
 *
 * @author Ruben Gees
 */
data class Notification(
    @Json(name = "id") override val id: String,
    @Json(name = "type") val type: NotificationType,
    @Json(name = "tid") val contentId: String,
    @Json(name = "link") val contentLink: HttpUrl,
    @Json(name = "linktext") val text: String,
    @Json(name = "time") override val date: Date,
    @Json(name = "description") val additionalDescription: String?
) : ProxerIdItem, ProxerDateItem
